src/re2parser.cc

```diff
@@ -0,0 +1,522 @@
+         * @param nfa vata2::Nfa::Nfa in which the states will be made final
+         */
+        void make_state_final(int state, Vata2::Nfa::Nfa &nfa) {
+            for (auto target_state: this->state_cache.state_mapping[state]) {
+                // States without an incoming edge should not be in the automata
+                if (!this->state_cache.has_state_incoming_edge[target_state]) {
+                    continue;
+                }
+                nfa.add_final(target_state);
+            }
+        }
+
+        /**
+         * Renumbers the states of the input_nfa to be from <0, numberOfStates>
+         * @param program_size Size of the RE2 prog
+         * @param input_nfa vata2::Nfa::Nfa which states should be renumbered
+         * @return Same vata2::Nfa::Nfa as input_nfa but with states from interval <0, numberOfStates>
+         */
+        static Vata2::Nfa::Nfa renumber_states(int program_size, Vata2::Nfa::Nfa &input_nfa) {
+            std::vector<unsigned long> renumbered_states(program_size, -1);
+            Vata2::Nfa::Nfa renumbered_explicit_nfa(0);
+            for (int state = 0; state < program_size; state++) {
+                const auto& transition_list = input_nfa.get_transitions_from_state(state);
+                // If the transition list is empty, the state is not used
+                if (transition_list.empty()) {
+                    continue;
+                } else {
+                    // addNewState returns next unused state of the new NFA, so we map it to the original state
+                    renumbered_states[state] = renumbered_explicit_nfa.add_new_state();
+                }
+            }
+
+            for (auto state: input_nfa.finalstates) {
+                if (static_cast<int>(renumbered_states[state]) == -1) {
+                    renumbered_states[state] = renumbered_explicit_nfa.add_new_state();
+                }
+                renumbered_explicit_nfa.add_final(renumbered_states[state]);
+            }
+
+            for (int state = 0; state < program_size; state++) {
+                const auto& transition_list = input_nfa.get_transitions_from_state(state);
+                for (const auto& transition: transition_list) {
+                    for (auto stateTo: transition.states_to) {
+                        renumbered_explicit_nfa.add_trans(renumbered_states[state], transition.symbol,
+                                                          renumbered_states[stateTo]);
+                    }
+                }
+            }
+
+            for (auto state: input_nfa.initialstates) {
+                renumbered_explicit_nfa.add_initial(renumbered_states[state]);
+            }
+
+            return renumbered_explicit_nfa;
+        }
+        };
+}
+
+ /**
+ * The main method, it creates NFA from regex
+ * @param pattern regex as string
+ * @return vata2::Nfa::Nfa corresponding to pattern
+ */
+Vata2::Nfa::Nfa Vata2::RE2Parser::create_nfa(const std::string& pattern) {
```

Comment:
I would suggest to create paired function with signature `void create_nfa(Nfa*, string)`